File: leetcode_problems/cpp/lc112_path_sum.cpp
#include <algorithm>

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution
{
public:
    bool countPathSum(TreeNode *node, int sum, int target)
    {
        if (node == nullptr)
        {
            return false;
        }
        sum += node->val;
        if (sum == target && node->left == nullptr && node->right == nullptr)
        {
            return true;
        }
        return countPathSum(node->left, sum, target) || countPathSum(node->right, sum, target);
    }

    bool hasPathSum(TreeNode *root, int targetSum)
    {
        if (root == nullptr)
        {
            return false;
        }
        return countPathSum(root, 0, targetSum);
    }
};

File: leetcode_problems/typescript/src/lc112_path_sum.ts
function hasPathSumRec(
  node: TreeNode | null,
  sum: number,
  targetSum: number
): boolean {
  if (node === null) {
    return false;
  }
  sum += node.val;
  if (sum === targetSum && node.left === null && node.right === null) {
    return true;
  }
  return (
    hasPathSumRec(node.left, sum, targetSum) ||
    hasPathSumRec(node.right, sum, targetSum)
  );
}

function hasPathSum(root: TreeNode | null, targetSum: number): boolean {
  if (root === null) {
    return false;
  }
  return hasPathSumRec(root, 0, targetSum);
}
